src/main/java/com/google/gcp_variant_transforms/library/VariantToBqUtils.java

```diff
@@ -0,0 +1,224 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility methods for write values with defined value type into Big Query rows.
+ */
+public class VariantToBqUtils {
+  public static String addReferenceBase(VariantContext variantContext) {
+    Allele referenceAllele = variantContext.getReference();
+    // If the ref length is longer than 1, store the ref as a String
+    String referenceBase = referenceAllele.getDisplayString();
+    return referenceBase.equals(Constants.MISSING_FIELD_VALUE) ? null : referenceBase;
+  }
+
+  public static String addNames(VariantContext variantContext) {
+    // If the ID field is ".", should write null into BQ row
+    String name = variantContext.getID();
+    return name.equals(Constants.MISSING_FIELD_VALUE) ? null : name;
+  }
+
+  public static List<TableRow> addAlternates(VariantContext variantContext,
+                                             Map<String, Integer> alleleIndexingMap) {
+    List<TableRow> altMetadata = new ArrayList<>();
+    List<Allele> altAlleles = variantContext.getAlternateAlleles();
+    // if field value is '.', alternateAllelesList will be empty
+    if (altAlleles.isEmpty()) {
+      TableRow subRow = new TableRow();
+      subRow.set(Constants.ColumnKeyConstants.ALTERNATE_BASES_ALT, null);
+      altMetadata.add(subRow);
+    } else {
+      for (int i = 0; i < altAlleles.size(); ++i) {
+        TableRow subRow = new TableRow();
+        String altBase = altAlleles.get(i).getDisplayString();
+        alleleIndexingMap.put(altBase, i + 1);  // the alt allele index is 1-based
+        subRow.set(Constants.ColumnKeyConstants.ALTERNATE_BASES_ALT, altBase);
+        altMetadata.add(subRow);
+      }
+    }
+    return altMetadata;
+  }
+
+  public static Set<String> addFilters(VariantContext variantContext) {
+    Set<String> filters = variantContext.getFilters();
+    if (filters.isEmpty()) {
+      filters = new HashSet<>();
+      filters.add("PASS");
+    }
+    return filters;
+  }
+
+  /**
+   * Add variant Info field into BQ table row.
+   * If the info field number in the VCF header is `A`, add it to the ALT sub field.
+   */
+  public static void addInfo(TableRow row, VariantContext variantContext,
+                             List<TableRow> altMetadata, VCFHeader vcfHeader) {
+    for (Map.Entry<String, Object> entry : variantContext.getAttributes().entrySet()) {
+      String attrName = entry.getKey();
+      Object value = entry.getValue();
+      VCFInfoHeaderLine infoMetadata = vcfHeader.getInfoHeaderLine(attrName);
+      VCFHeaderLineType infoType = infoMetadata.getType();
+      VCFHeaderLineCount infoCount = infoMetadata.getCountType();
+      if (infoCount == VCFHeaderLineCount.A) {
+        // put this info into ALT field
+        splitAlternateAlleleInfoFields(attrName, value, altMetadata, infoType);
+      } else {
+        row.set(attrName, convertToDefinedType(value, infoType));
```

Comment:
I am not still sure if we need to implement the logic of converting to type ouerselves. I think it's much safer is we rely on HTSJDK's implementation:
* `getAttributeAsBoolean`
* `getAttributeAsDouble`
* `getAttributeAsDoubleList`
* `getAttributeAsInt`
* `getAttributeAsIntList`
* `getAttributeAsString` 
* `getAttributeAsStringList`

All we need a swtich where calls the corresponding method based on the `infoType` value.

Please let me know what you think.

Reply:
getAttributeAsInt will return int instead of Integer. We cannot set default value as null. Moreover, those functions only cast the values into target type, like (Integer)object, I do not think it's a safer implementation